### ISSUE TITLE
Question: how to set apiUrl after module init

### ISSUE THREAD

**Issue by mbenedettini** [ng2-ui/map#271]
I'm retrieving my Google Maps key from server side (using a service), so I cannot call `NguiMapModule.forRoot()` with the right apiUrl from within my module imports. 

How can that be accomplished? Something similar was asked in this issue https://github.com/ng2-ui/map/issues/242 where user wants to change the language but I cannot figure out how to set apiUrl once the module has been initialized.

To sum it up, I'd like to set apiUrl from within my component constructor.

Could you guys point me in the right direction?
Thanks!

**Comment by aitboudad** [comment]
can try with:
```ts
  constructor(@Inject(NG_MAP_CONFIG_TOKEN) config) {}
  changeUrl() {
    config.apiUrl = 'your_custom_url';
  }
```

**Comment by mbenedettini** [comment]
It worked, thanks! I'm leaving some details about my implementation in case someone else comes across the same:

Here's my custom service which extends `NgMapAsyncCallbackApiLoader` introducing `changeUrl()`:
```javascript
//myngmap-apiloader-service.ts
import { Injectable } from '@angular/core';
import { NgMapAsyncCallbackApiLoader } from '@ngui/map';

@Injectable()
export class MyNgMapApiLoader extends NgMapAsyncCallbackApiLoader {
    changeUrl(url) {
        this.config.apiUrl = url;
    }
}
```

In my module I've added the following lines to the `providers` section:

```javascript
providers: [
        MyNgMapApiLoader,
        {
            provide: NgMapApiLoader,
            useClass: MyNgMapApiLoader
        }
    ]
```

And then I was able to change the url from within my component where ngmap is being used:

```javascript
// map.component.ts

export class MapComponent implements OnInit {
  constructor(
    private configService: ConfigService,
    private mapApiLoader: MyNgMapApiLoader
  ) {
    const GoogleMapsKey = configService.current.GoogleMapsKey;
    const url = `https://maps.google.com/maps/api/js?\
libraries=drawing,places,visualization&key=${GoogleMapsKey}`;
    mapApiLoader.changeUrl(url);
  }
}
```

**Comment by redmission84** [comment]
@mbenedettini I can't thank you enough for this! Worked perfectly.  I was struggling as we have seperate keys for our dev environments to run in localhost and prod, and wanted to load the API key from a config file. You saved me. :)

**Comment by inzi** [comment]
I'm new to this - do you have a get repo with this in it I could look at? 

This is very similar to what I'm trying to accomplish.

Thanks in advance.

**Comment by mbenedettini** [comment]
@inzi sorry, I have it in a private project. I can give you some more insights:

+ The first snippet is a new service you need to introduce into your app. This service will later be used in your component where you want to use ng2-ui/map .
+ The 2nd snippet is what I had to add in the providers section of my module
+ Finally, the 3rd snippet shows how to change the api url by calling `mapApiLoader.changeUrl(url);`

If you can create a repo with a minimum set up (just a simple component displaying a map) I'd be happy to help you.

**Comment by inzi** [comment]
Thank you for the kind offer.

I inherited an angular2 project I am trying to maintain, and I keep having to rebuild and deploy every time a different api key is used, so I feel your pain. I'm still learning about bootstrapping my app using a root module, etc.

So I guess a first question is if this is even possible with Angular2, or must I convert this to an Angular 6 project, first?  If this isn't possible in Angular2, then maybe I should focus on upgrading it to angular6, then start from there.

I've obviously got more learning to do.

**Comment by mbenedettini** [comment]
You're going to have to convert it to Angular 6 sooner or later. I'd rather do it as soon as possible.

**Comment by inzi** [comment]
Agreed.

I was able to get it to work! Thank you for your tips. They put me on the right path.

I created the service the same as you - for anyone else that's a total n00b like me:

In your main app module, you still need to pull in NguiMapModule.
Like so:
    NguiMapModule.forRoot({ apiUrl: 'https://maps.google.com/maps/api/js?key=&libraries=visualization,places,drawing' }),

Not I just left the key out.

In your main app module, add the provider above.

After you bootstrap your app and get your API key, on the component for your page, you inject it there, and set the URL in the constructor. 

In my case, I did a normal bootstrap to pull down a JSON string and set it to a constants service.

Then, I referenced the constants service, which constructed my api url, and set it.
Like so:

  constructor(
    injector: Injector,
    private _MapLoader: MyNgMapApiLoader,
    private ref: ChangeDetectorRef
  ) {
    super(injector);
    this._MapLoader.changeUrl(AppConsts.googleMapsURL());
  }
